### PR TITLE
Dry Run Task API

### PR DESCRIPTION
**Pull request by mkam**
Adds the API endpoint for inspecting a task without actually creating the task. Will eventually be used by the create task CLI to output the plan before prompting user approval.

The task is initialized, an inspection is run, and the plan of that inspection is returned in the response. The task is never added to the map of drivers, though, so it is essentially discarded at the end of the inspection.

I left the working directory that was created during initialization on the off chance a user has their working directory set to the wrong path and we delete something unintentionally. There might be a better way to handle this, but I think that could always be a follow up change.

**Example**
POST `/v1/dryrun_tasks`
Request
```
{
    "description": "Writes the service name, id, and IP address to a file",
    "enabled": true,
    "name": "inspected_task",
    "providers": [
        "local"
    ],
    "services": [
        "api"
    ],
    "source": "mkam/hello/cts"
}
```
Response
200 OK
```
{
    "request_id": "d1942c0b-04ef-cd4d-70c4-f507f2bc9538",
    "run": {
        "plan": "\nTerraform used the selected providers to generate the following execution\nplan. Resource actions are indicated with the following symbols:\n  + create\n\nTerraform will perform the following actions:\n\n  # module.inspected_task.local_file.greeting will be created\n  + resource \"local_file\" \"greeting\" {\n      + content              = \"Hello!\"\n      + directory_permission = \"0777\"\n      + file_permission      = \"0777\"\n      + filename             = \"greetings.txt\"\n      + id                   = (known after apply)\n    }\n\n  # module.inspected_task.local_file.greeting_services will be created\n  + resource \"local_file\" \"greeting_services\" {\n      + content              = <<-EOT\n            Hello, api!\n        EOT\n      + directory_permission = \"0777\"\n      + file_permission      = \"0777\"\n      + filename             = \"services_greetings.txt\"\n      + id                   = (known after apply)\n    }\n\nPlan: 2 to add, 0 to change, 0 to destroy.\n"
    },
    "task": {
        "buffer_period": {
            "enabled": true,
            "max": "20s",
            "min": "5s"
        },
        "condition": {},
        "description": "Writes the service name, id, and IP address to a file",
        "enabled": true,
        "name": "inspected_task",
        "providers": [
            "local"
        ],
        "services": [
            "api"
        ],
        "source": "mkam/hello/cts",
        "source_input": {},
        "version": "",
        "working_dir": "sync-tasks/inspected_task"
    }
}
```

Changes:
- Open API spec changes for naming the endpoint and removing the unused run response attributes
- Implementing the API code for running an inspect on a task without creating it
- Returning the Terraform plan output on an InspectPlan
- Updating the default error code from 500 to 400

Subtask of https://github.com/hashicorp/consul-terraform-sync/issues/522

### CODE REVIEW

**Review by lornasong — COMMENTED** [review]
Looks good to me! Testing looks thorough and changes easy to follow.

One question I was wondering - wanted to catch up on your latest thoughts on the endpoint path. One particular thing that I wanted to hear your thoughts on is the terminology switch from "dry run" to "inspect". No objects from me, just wanted to learn the thought process

Edit: I forgot to click approve so accidentally defaulted to just comment - apologies for any confusion

**Review by lornasong — APPROVED** [review]

**Comment by mkam** [comment]
@lornasong Totally valid question, I've been wrestling with what to call this stuff! And I'm still not sure if I went with the best option to be honest! But I figured a name change would be easy to make after the implementation is completed, and I didn't want to block starting the code anymore than I already had.

I was trying to keep REST practices in mind and wanted to keep the endpoint a noun, which is why I didn't go with something like `/inspect_task`. I was also trying to convey that these are still new tasks, just not persisted tasks. So with this thinking, using dry run as a descriptor for tasks made sense to me. But I get what you're saying, it's not a term we've used before, maybe `/inspected_tasks` would have been better?

File: api/openapi.yaml [hunk 1]
```diff
@@ -88,10 +88,10 @@ paths:
               schema:
                 $ref: '#/components/schemas/ErrorResponse'
 
-  /v1/dryrun/task:
+  /v1/dryrun_tasks:
     post:
       summary: Triggers a dryrun for a provided task
-      operationId: executeTaskDryrun
+      operationId: createDryRunTask
```
[the remark this answers is not in the document]
Comment (mkam) [reply api/openapi.yaml:94]: Sure! To me, the endpoint and the handler is for all dry run tasks, but a POST to this endpoint specifically represents creating a single task. The `/tasks` endpoint might be a better example, where a POST to `/tasks` creates a single task, but a GET on `/tasks` would return a list of all tasks.

[the remark this answers is not in the document]
Comment (lornasong) [reply api/openapi.yaml:94]: Thanks for sharing about how to orient around thinking about the task/s - that makes sense!

**Comment by lornasong** [comment]
Thanks for sharing! Completely agree with your strategy to move ahead with implementing and circling-back if there's a name change. The naming isn't a blocker, so feel free to merge and we can continue to chat.

Good point about keeping the path a noun. "Inspect" is hard to noun-ify 😆 . `inspected_tasks` sounds as though it could mean a previously inspected task so I think it could be potentially a little confusing.

I just searched "dry run api". It looks like choosing between `/<dry-run>-<noun>` vs. `/noun?<dry-run-query-param>` is a common question (stack overflow). At the top of the results, I saw Google & Cisco's APIs go with the query param. I think earlier there were concerns about the code path for `?inspect=true` - has that changed at all as this is implemented or maybe still the same?

**Comment by mkam** [comment]
I believe we went away from something like `/tasks?inspect=true` is because the behavior when this parameter was set was so different than the standard create. The response code was different (201 vs 200), the response body was different (the terraform plan is returned in addition to the task), and just the general behavior of no resource is created. But could be worth reevaluating if there are examples of that not being an issue in the wild!

**Comment by lornasong** [comment]
Thanks for the details! I see, a lot of the user-facing return is different. Yea, the few examples I came across used a query parameter. Briefly looking into some docs I didn't see info on how they handled the response status and body (I'm assuming the behavior is different). Agree that it could be worth researching a little more to see what the industry is doing

**Pull request merged**